audio/out/ao_aaudio.c

```diff
@@ -0,0 +1,454 @@
+                mp_chmap_sel_add_map(&sel, &aaudio_chmaps[i]);
+
+            if (!ao_chmap_sel_adjust(ao, &sel, &ao->channels)) {
+                MP_ERR(ao, "Failed to find channel map");
+                goto err;
+            }
+
+            for (uint8_t i = 0; i < MP_ARRAY_SIZE(aaudio_chmaps); i++) {
+                if (mp_chmap_equals(&aaudio_chmaps[i], &ao->channels)) {
+                    channel_mask = aaudio_masks[i];
+                    break;
+                }
+            }
+            assert(channel_mask != AAUDIO_CHANNEL_INVALID);
+
+            p->AAudioStreamBuilder_setChannelMask(builder, channel_mask);
+        } else {
+            p->AAudioStreamBuilder_setChannelCount(builder, ao->channels.num);
+        }
+
+        p->AAudioStreamBuilder_setDeviceId(builder, p->device_id);
+        p->AAudioStreamBuilder_setDirection(builder, AAUDIO_DIRECTION_OUTPUT);
+        p->AAudioStreamBuilder_setSharingMode(builder,
+                                              (ao->init_flags & AO_INIT_EXCLUSIVE) ?
+                                                  AAUDIO_SHARING_MODE_EXCLUSIVE :
+                                                  AAUDIO_SHARING_MODE_SHARED);
+        p->AAudioStreamBuilder_setFormat(builder, format);
+        p->AAudioStreamBuilder_setSampleRate(builder, ao->samplerate);
+        p->AAudioStreamBuilder_setErrorCallback(builder, error_callback, ao);
+        p->AAudioStreamBuilder_setBufferCapacityInFrames(builder, p->buffer_capacity);
+        p->AAudioStreamBuilder_setPerformanceMode(builder, p->performance_mode);
+        p->AAudioStreamBuilder_setDataCallback(builder, data_callback, ao);
+
+        if (p->device_api >= 28) {
+            p->AAudioStreamBuilder_setContentType(builder, (ao->init_flags &
+                                                            AO_INIT_MEDIA_ROLE_MUSIC) ?
+                                                               AAUDIO_CONTENT_TYPE_MUSIC :
+                                                               AAUDIO_CONTENT_TYPE_MOVIE);
+            p->AAudioStreamBuilder_setUsage(builder, AAUDIO_USAGE_MEDIA);
+        }
+
+        p->builder = builder;
+
+        if ((result = p->AAudioStreamBuilder_openStream(p->builder, &p->stream)) >= 0) {
+            if (p->device_api < 32) {
+                int32_t channel_count = p->AAudioStream_getChannelCount(p->stream);
+
+                if (channel_count < 0 ||
+                    channel_count >= MP_ARRAY_SIZE(aaudio_default_chmaps) ||
+                    (ao->channels = aaudio_default_chmaps[channel_count]).num == 0) {
+                    MP_ERR(ao, "Unknown layout for channel count: %" PRId32 "\n", channel_count);
+
+                    p->AAudioStream_close(p->stream);
+                    p->stream = NULL;
+                    goto err;
+                }
+            }
+
+            ao->device_buffer = p->AAudioStream_getBufferCapacityInFrames(p->stream);
+            return 1;
```

Comment:
might be personal preference, but: the success path shouldn't be deeply nested. use early returns.